text/0000-parent-trait-use.md

```diff
@@ -0,0 +1,176 @@
+You see, if you specify `Output` in all parent traits in angle brackets, the `Output` type is never used and you can even specify it yourself.
+
+Now you are the maintainer of a crate. You defined a trait `A`, which everyone uses, that looks like this:
+```
+trait A {
+    fn a();
+    fn b();
+    fn c();
+}
+```
+You recognize, it may be useful to split parts of this traits into single traits with less functionality. You split some of the functions into seperate traits, which looks as the first example:
+```
+trait B {
+    fn b();
+}
+trait C {
+    fn c();
+}
+trait A: B+C {
+    fn a();
+}
+```
+Now every user, that implemneted `A`, now has to implement `B` and `C`, so it would break their crates when updating version.
+The solution is the same as before:
+```
+trait A: use B+use C {
+    fn a();
+}
+```
+
+In order not to break anything, this change has to be recursive.
+Someone may already use `A` as a using parent trait:
+```
+trait X: use A {}
+```
+This trait will need to use all functions, that are impmenetable inside `A`, so after the split of `A` it will also use functions of `B` and `C`.
+
+This use case is probably more important than ergonomics.
+
+When defining a trait, that uses it's parent traits, it will just expand to multiple trait definitions. If a trait is used, and at least one of it's functions is defined in this trait, it will define the trait. If multiple traits use the same trait, only one of them is allowed to define functions of that trait.
+If types are defined in a supertrait, but are set using angle brackets, these types will be implictely set to the type in angle brackets after expanding.
+
+Assuming you want to implement the previously mentioned `Math` trait for some type, you could write this:
+```
+type Number;
+impl Math for Number {
+    type Output = Number;
+    fn add(self, rhs: Self) -> Self {…}
+    fn sub(self, rhs: Self) -> Self {…}
+}
+```
+This would expand to following:
+```
+type Number;
+impl Add for Number {
+    type Output = Number;
+    fn add(self, rhs: Self) -> Self {…}
+}
+impl Math for Number {
+    type Output = Number;
+    fn sub(self, rhs: Self) -> Self {…}
+}
+impl Math for Number {
+    type Output = Number;
+}
+```
+You would still need to implement `Mul` and `Div`, else you will get errors, as if you wrote the expandet form directly. It's also possible to implement these traits seperately.
+
+This won't affect existing code. It may be still preferable to define everything inside the traits, they are used as. Implementing a trait, that uses it's parent traits should only be used for important types.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This will be implemented by recursively searching for all names of traits, which are used by the current trait.
+If one of the names exists at least twice, this will be reported as an error when the trait is defined.
```

Comment:
Currently, it's a backwards compatible change to add default methods to a trait. This would cause that to break, since `trait Foo: use Bar { fn foo() { ... } }` would break if `Bar` added a default method called `foo`.

Reply:
Okay, then this may be a bit more complicated, than I thought.
Then my other approach for using something as defined in `Alternatives` seems more useful:
```
trait Bar {
    fn bar();
    fn foo() {…}
}

trait Foo: Bar {
    use Bar::bar; //now bar is implementable when implementing `Foo`
    fn foo();
}
```
Then another problem occurs: It's not possible to use traits, when the user doesn't have to implement one of it's functions, but this was not possible anyway, according to my descriptions.
The child trait is required to use everything of the parent trait, that has no default value. This way, it will be ensured, that adding something with default impl won't be a breaking change now.